### PR TITLE
Distinguish between DML and the generic 'GPU' term. This is needed for packaging DML EP in the same ORT GPU pkg.

### PR DESCRIPTION
**Pull request by pranavsharma**
### Description
We want to package DML EP in the same pkg as ORT GPU pkg. This is one of the changes required to do so. Other 2 changes are
- packaging changes (https://github.com/microsoft/onnxruntime/pull/22395)
- delay loading changes in python binding (@snnn)

### Motivation and Context
Users want DML, TensorRT and CUDA EP in the same ORT GPU pkg so that they can easily switch between them.

### CODE REVIEW

**Review by snnn — APPROVED** [review]

File: include/onnxruntime/core/framework/ortdevice.h [hunk 1]
```diff
@@ -17,6 +17,7 @@ struct OrtDevice {
   static const DeviceType GPU = 1;  // Nvidia or AMD
```
Comment (fdwr) [inline include/onnxruntime/core/framework/ortdevice.h:17]: Would it be clearer (but possibly a much large search and replace, and so I understand the hesitation) to rename this to CUDA?

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 2]
```diff
@@ -288,11 +288,9 @@ const char* GetDeviceName(const OrtDevice& device) {
     case OrtDevice::CPU:
       return CPU;
     case OrtDevice::GPU:
-#ifdef USE_DML
-      return DML;
-#else
       return CUDA;
-#endif
+    case OrtDevice::DML:
+      return DML;
```
Comment (fdwr) [inline onnxruntime/python/onnxruntime_pybind_state.cc:293]: I see parts of the DML EP use `OrtDevice::GPU`, and so I wonder the impact of splitting this out; but it appears this change would only impact Python callers at most (not existing C++ callers).

```
    BucketizedBufferAllocator::BucketizedBufferAllocator(
        ID3D12Device* device,
        ExecutionContext* context,
        const D3D12_HEAP_PROPERTIES& heapProps,
        D3D12_HEAP_FLAGS heapFlags,
        D3D12_RESOURCE_FLAGS resourceFlags,
        D3D12_RESOURCE_STATES initialState,
        std::unique_ptr<DmlSubAllocator>&& subAllocator
        )
        : onnxruntime::IAllocator(
            OrtMemoryInfo(
                "DML",
                OrtAllocatorType::OrtDeviceAllocator,
                OrtDevice(OrtDevice::GPU, OrtDevice::MemType::DEFAULT, 0) <<<<<<<<<<<<<<<<<
            )
        ),
```

```
    ExecutionProvider::ExecutionProvider(
        IDMLDevice* dmlDevice,
        Dml::ExecutionContext* executionContext,
        bool enableMetacommands,
        bool enableGraphCapture,
        bool enableSyncSpinning,
        bool disableMemoryArena) :
            IExecutionProvider(onnxruntime::kDmlExecutionProvider, OrtDevice(OrtDevice::GPU, OrtDevice::MemType::DEFAULT, 0))  <<<<<<<<<<<<<<<<<
```

**Review by fdwr — APPROVED** [review]
Looks okay to me. FYI @sumitsays.

File: include/onnxruntime/core/framework/ortdevice.h (hunk 1 above)
Comment (pranavsharma) [reply include/onnxruntime/core/framework/ortdevice.h:17]: Yeah, trying not to boil the ocean for this change. Your comment is well taken though and something that needs cleanup.

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 2 above)
Comment (pranavsharma) [reply onnxruntime/python/onnxruntime_pybind_state.cc:293]: Thanks for pointing this out. Fixed in the next commit. It's good to keep things consistent.

Comment (fdwr) [reply onnxruntime/python/onnxruntime_pybind_state.cc:293]: I just checked my little DML sample (https://github.com/fdwr/OnnxRuntimeDirectMLEPSample), and I don't explicitly refer to the `OrtMemoryInfoDeviceType` anywhere anyway, just memoryInformationName ("DML", "Cpu"), `OrtAllocatorType` (`OrtDeviceAllocator`), and `OrtMemType` (`OrtMemTypeDefault`).

https://github.com/microsoft/onnxruntime/blob/b4afc6266f7ff20e7b79eaea7fa62f3e30b7474f/include/onnxruntime/core/session/onnxruntime_c_api.h#L393

It's interesting the difference between this enum set and the public enum:

```
typedef enum OrtMemoryInfoDeviceType {
  OrtMemoryInfoDeviceType_CPU = 0,
  OrtMemoryInfoDeviceType_GPU = 1,
  OrtMemoryInfoDeviceType_FPGA = 2
} OrtMemoryInfoDeviceType;
```

**Review by fdwr — DISMISSED** [review]
👍 Sumit/Patrice, do you have any concerns?

**Comment by pranavsharma** [comment]
> 👍 Sumit/Patrice, do you have any concerns?

Looks like there are some DML test failures. I'm investigating. @sumitsays/ @fdwr / @PatriceVignola - if something looks obvious, let me know. thanks!

**Comment by tianleiwu** [comment]
Do we plan to support using DML and CUDA EP in same session? If so, some part like gpu data transfer need support transfer tensors between DML and GPU memory. If not, we might need add some check to prevent the usage.

**Comment by snnn** [comment]
> Do we plan to support using DML and CUDA EP in same session? If so, some part like gpu data transfer need support transfer tensors between DML and GPU memory. If not, we might need add some check to prevent the usage.

No. They cannot be both enabled in the same process.

**Pull request closed without merging**